### ISSUE TITLE
Linux: Cpu Platforms

### ISSUE THREAD

**Issue by borgdylan** [dotnet/coreclr#60]
I realize that currently only Linux on 64 bit is supported. Will Linux on 32 bit x86 and on ARM be supported as well?

**Comment by sharique** [comment]
I don't think supporting 32-bit is worth effort, as you won't find new 32-bit hardware. I guess ARM should be in roadmap. 
I think Mono framework already support ARM.

**Comment by borgdylan** [comment]
Mono also supports 32-bit linux. Running 32 bit linux on new hardware gives memory savings plus its required if developing for certain Intel IoT boards.

**Comment by n8ohu** [comment]
Indeed; none of my current Linux machines run a 64 bit version of Linux, primarily because I utilize numerous applications that would require complex workarounds to run on 64 bit.

**Comment by sergiy-k** [comment]
Yes, right now we support 64bit only. There is still a lot of work ahead before we can say that it is ready. Once it is functional we can consider other platforms as well. But for now 64bit is our main focus.

**Comment by OtherCrashOverride** [comment]
32bit support is a must.  Even though hardware may be 64bit, I don't see a lot of on host deployments.  I think the main use case is virtual machines where light weight deployment is preferred to a full and heavy IIS/ASP.Net Windows server install.

My intended use case is 32bit x86 virtual machines and 32bit ARM SoC devices (IoT).  For 'heavy' deployments, I already have Mono.  I am thrilled at the 'just the basics' approach of .Net Core.

**Comment by OtherCrashOverride** [comment]
I should also point out that the 'deal breaker' comes down to remote debugging support with Visual Studio.  Without this feature, .Net Core is useless to me regardless of 32/64 bit operation.

**Comment by sergiy-k** [comment]
We are working on debugging tools too. We have started bringing up CoreCLR on Linux only recently so there is still a lot of rough edges and missing pieces. Please bear with us. :)

**Comment by borgdylan** [comment]
@jkotas is this marked as up-for-grabs as you will accept PRs for it? In that case I might try to infer a linux 32bit port from the 64bit port.

**Comment by jkotas** [comment]
Yes. I have just added a note about it to https://github.com/dotnet/coreclr/wiki/Contribution-guidelines

We are most likely going to accept any good changes to bring CoreCLR to new platforms and operating systems.

**Comment by jkotas** [comment]
We prefer to get smaller incremental PR for this kind of work, e.g. make the PAL build, port all .asm files, etc.

**Comment by borgdylan** [comment]
Would there be a recommended order of tasks i.e the steps to do a port. Since I have invested in 32 bit Linux (with Physical Address Extension) I would like to try to do the coreclr port to it if i have what it takes.

**Comment by borgdylan** [comment]
@jkotas I tried to enable the 32bit build on linux but I am getting many C++ compiler errors. Can anyone guide me on how to proceed? My cmakelists changes are here https://github.com/borgdylan/coreclr/commit/75c5769a8a7496d38644a02a909c3270ad03f148

**Comment by jkotas** [comment]
The initial work to port to a new target is about fixing the compile errors. Check the changes in #117 - majority of them are just dealing with compile errors. It is normal that there is a lot of them when you run build for the first them. Once you start understanding and fixing them, you will discover that there are recurring patterns and so it is not that bad after all.

**Comment by borgdylan** [comment]
One of the repeating erros is "error: 'operator new' takes type size_t ('unsigned int') as first parameter
    void *operator new(size_t size)". The compiler does not say why taking in size_t is bad here.

**Comment by sergiy-k** [comment]
@borgdylan If you would like to do this work I can certainly help you with that and guide you through the process. If yes, then let’s split the work into smaller work items. I can open a separate ‘up-for-grabs’ issue for each of them. For example, I were to do this work I would stage it in the following way:

1. Add support for 32bit build in CMake files for Linux
2. Make PAL compile in x86 mode on Linux. This can be split into more detailed work items since PAL consists of different components and maybe someone else can help you to do them in parallel. Personally I would start by first making the files under coreclr\src\pal\src\init compile on 32bit Linux. Once they compile the rest of the PAL files will be much easier to port (you can enable one PAL component/folder at a time in coreclr\src\pal\src\CMakeLists.txt).
3. Make PAL tests (coreclr\src\pal\tests) compile for x86 on Linux.
4. Fix all failing PAL tests.
5. Make CoreCLR Runtime/JIT components compile for x86. For example, this includes the utilcode, gcinfo, vm, jit, etc. Implement x86 version of assembly stubs (we will need to port all *.S files that can be found under coreclr\src\vm\amd64). Again we can create separate ‘up-for-grabs’ issues for each of this components.
6. Build x86 version of libcoreclr.so (coreclr\src\dlls\mscoree\coreclr\CMakeLists.txt).
7. Build x86 version of corerun for Linux (coreclr\src\coreclr\hosts\unixcorerun\CMakeLists.txt)
8. Now the fun part begins: bring up the x86 version of runtime (debug thorough startup issues) and make HelloWorld run. Given the current state of the system, it should not be too hard.

Please let me know what you think. Thank you.

**Comment by sergiy-k** [comment]
@borgdylan Could you post full description of the error you are seeing so I could see file name and line number where it is coming from? It also would be helpful if you could show output of 'make VERBOSE=1' for that file in order to understand which defines are enabled during compilation.

**Comment by borgdylan** [comment]
OK.I would like to try doing the port.

**Comment by borgdylan** [comment]
One of the erros is:

```
In file included from /var/www/git/coreclr/src/vm/classhash.cpp:11:
In file included from /var/www/git/coreclr/src/vm/common.h:301:
In file included from /var/www/git/coreclr/src/vm/ibclogger.h:19:
/var/www/git/coreclr/src/vm/crst.h:434:11: error: 'operator new' takes type
      size_t ('unsigned int') as first parameter
    void *operator new(size_t size)
          ^
```

In my opinion the error is just stating what is in a header file but not why it is wrong.

**Comment by borgdylan** [comment]
The defines are:
```
-DBETA=0 -DBIT32=1 -DBUILDENV_CHECKED=1 -DCONDITION_HANDLING=1 -DDBG_TARGET_I386 -DDBG_TARGET_I386_UNIX -DDBG_TARGET_X86 -DDBG_TARGET_X86_UNIX -DDEBUG -DDEBUGGING_SUPPORTED -DDEV10 -DDEVL=1 -DDISABLE_CONTRACTS -DFEATURE_APPDOMAIN_RESOURCE_MONITORING -DFEATURE_ARRAYSTUB_AS_IL -DFEATURE_ASYNC_IO -DFEATURE_BCL_FORMATTING -DFEATURE_COLLECTIBLE_TYPES -DFEATURE_CORECLR -DFEATURE_CORESYSTEM -DFEATURE_CORRUPTING_EXCEPTIONS -DFEATURE_EXCEPTIONDISPATCHINFO -DFEATURE_FRAMEWORK_INTERNAL -DFEATURE_HOSTED_BINDER -DFEATURE_HOST_ASSEMBLY_RESOLVER -DFEATURE_IMPLICIT_TLS -DFEATURE_ISYM_READER -DFEATURE_LEGACYNETCF -DFEATURE_LEGACYNETCFFAS -DFEATURE_LEGACYNETCF_DBG_HOST_CONTROL -DFEATURE_LOADER_OPTIMIZATION -DFEATURE_MAIN_CLR_MODULE_USES_CORE_NAME -DFEATURE_MANAGED_ETW -DFEATURE_MANAGED_ETW_CHANNELS -DFEATURE_MERGE_CULTURE_SUPPORT_AND_ENGINE -DFEATURE_MULTICOREJIT -DFEATURE_NORM_IDNA_ONLY -DFEATURE_PAL -DFEATURE_PAL_SXS -DFEATURE_PREJIT -DFEATURE_STANDALONE_SN -DFEATURE_STRONGNAME_DELAY_SIGNING_ALLOWED -DFEATURE_STRONGNAME_MIGRATION -DFEATURE_STUBS_AS_IL -DFEATURE_SVR_GC -DFEATURE_SYMDIFF -DFEATURE_SYNTHETIC_CULTURES -DFEATURE_USE_ASM_GC_WRITE_BARRIERS -DFEATURE_VERSIONING -DFEATURE_WIN32_REGISTRY -DFEATURE_WINDOWSPHONE -DFEATURE_WINMD_RESILIENT -DFX_BRANCH_SYNC_COUNTER_VALUE=0 -DI386 -DLINUX32 -DNO_CRT -DNTDDI_VERSION=NTDDI_WIN8 -DNTMAKEENV -DNT_INST=0 -DOFFICIAL_BUILD=0 -DPLATFORM_UNIX=1 -DUNICODE -DUNIX_I386_ABI -DUNIX_X86_ABI -DURTBLDENV_FRIENDLY=Checked -DWIN32 -DWIN32_LEAN_AND_MEAN=1 -DWINNT=1 -DWINVER=0x0602 -DWRITE_BARRIER_CHECK=1 -DX86 -D_BLD_CLR -D_CRT_STDIO_ARBITRARY_WIDE_SPECIFIERS -D_DBG -D_DEBUG -D_I386 -D_I386_ -D_I386_SIMULATOR_ -D_I386_WORKAROUND_ -D_NEW_SDK=1 -D_SECURE_SCL=0 -D_SKIP_IF_SIMULATOR_ -D_TARGET_AMD64_=0 -D_TARGET_I386_=1 -D_TARGET_X86_=1 -D_UNICODE -D_WIN32_IE=0x0900 -D_WIN32_WINNT=0x0602 -D_X86 -D_X86_ -D_X86_SIMULATOR_ -D_X86_SIMULATOR_PERF_ -D_X86_WORKAROUND_ -D__LINUX__=1 -D_i386_SIMULATOR_PERF_
```
Could some error arise as WIN32 is also being defined?

**Comment by borgdylan** [comment]
What should the value of PAL_CS_NATIVE_DATA_SIZE be on 32 bit Linux? And why is it 96 on 64bit Linux instead of 64?

**Comment by borgdylan** [comment]
More errors spotted:
* There are cases where the PAL defines structs that are also defined in other files. Should the second version not be defined if the PAL version is defined?
* There cases of errors due to missing WIN32 defines such as LPMEMORYSTATUSEX. Is there a header in the PAL which should be included that has these defines?

**Comment by sergiy-k** [comment]
@borgdylan  Could you please explain why you think PAL_CS_NATIVE_DATA_SIZE should be 64 instead of 96 on 64bit Linux (sizeof(PAL_CRITICAL_SECTION_NATIVE_DATA) returns 96)? You can check the value for 32bit Linux in the same way.

LPMEMORYSTATUSEX is defined in pal.h but it’s currently under “#if defined(_AMD64_)” that’s why you these errors.

Could you please provide examples of the structs that are defined in multiple places and cause problems?

The other thing that could be causing problems is that you have _TARGET_AMD64_ defined in your settings.

Thank you.

**Comment by borgdylan** [comment]
I did not know that it was the size of PAL_CRITICAL_SECTION_NATIVE_DATA, If you tell me the header it is defined in, I will compute its size with sizeof.

Should I add "|| defined(X86)" to pal.h conditional compiles?

Some of teh structs that get defined twice are dispatcher related. I will respond later with actual struct names.

I will check for TARGET_AMD64. In the main cmake file  I should be defining TARGET_X86.

**Comment by borgdylan** [comment]
Repeated defines include RUNTIME_FUNCTION_INDIRECT,_RUNTIME_FUNCTION,_RUNTIME_FUNCTION

**Comment by borgdylan** [comment]
So PAL_CRITICAL_SECTION_NATIVE_DATA is 76 bytes on 32 bit Linux as per the following code:
```
//progam to get the size of PAL_CRITICAL_SECTION_NATIVE_DATA
// on the current cpu/os combination

#include <pthread.h>
#include <stdio.h>

//copied from coreclr/src/pal/src/sync/cs.cpp
typedef struct _PAL_CRITICAL_SECTION_NATIVE_DATA 
{
    pthread_mutex_t mutex;
    pthread_cond_t condition;
    int iPredicate;
} PAL_CRITICAL_SECTION_NATIVE_DATA;

int main() {
    printf("%i\n", sizeof(PAL_CRITICAL_SECTION_NATIVE_DATA));
    return 0;
}
```

**Comment by borgdylan** [comment]
Conflict in printf variants:
The PAL defines many printf variants using char * notation
While another file "printfcpp.hpp" defines them using LPSTR notation

The compiler is saying that the two definitions are not matching up?
We all know that on UNIX LPSTR becomes char * and LPCSTR becomes const char *. Should the code in printfcpp.hpp have its types change or should the PAL  define LPSTR and PLCSTR ins terms of char * and const char *?

**Comment by borgdylan** [comment]
New operator:
i am also seeing errors related to newing up arrays using the nothrow variable (says teh function for the new operator was not found). It seems like the new.hpp file in PAL is not being effective. I can confirm that it is being compiled.

**Comment by jkotas** [comment]
>>> LPMEMORYSTATUSEX is defined in pal.h but it’s currently under “#if defined(AMD64)
You should remove this ifdef. MEMORYSTATUSEX and related functions should be on for all platforms.

>>> RUNTIME_FUNCTION_INDIRECT, etc.
These should be ifdefed out for x86 in the PAL. You should be picking up the definitions from src\inc\corcompile.h.

**Comment by jkotas** [comment]
> LPMEMORYSTATUSEX is defined in pal.h but it’s currently under “#if defined(AMD64)

You should remove this ifdef. MEMORYSTATUSEX and related functions should be on for all platforms.

> RUNTIME_FUNCTION_INDIRECT, etc.

These should be ifdefed out for x86 in the PAL. You should be picking up the definitions from src\inc\corcompile.h.

**Comment by borgdylan** [comment]
@jkotas will fix and push to my fork

**Comment by borgdylan** [comment]
I may have found a solution to problems like;
```
In file included from /var/www/git/coreclr/src/vm/ceeload.cpp:15:
In file included from /var/www/git/coreclr/src/vm/common.h:78:
In file included from /var/www/git/coreclr/src/pal/inc/rt/stdint.h:6:
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:145:22: error: 'operator new' takes type size_t ('unsigned int') as first parameter
inline void *__cdecl operator new(size_t, void *_P)
'''
Clang may not be liking parameters that are nameless. Does MS Visual C++ allow both cases?

**Comment by janvorli** [comment]
I believe the issue with size_t in the new is that we define size_t as 
typedef unsigned long size_t;
but in 32 bit mode, it needs to be defined as
typedef unsigned int size_t;

There will be other data types like this in the pal\_mstypes.h (e.g. SIZE_T, SSIZE_T, ptrdiff_t, ...)

**Comment by borgdylan** [comment]
Thanks, I did not know that the PAL overrode basic types like size_t.

**Comment by borgdylan** [comment]
Fixing the definition killed all the errors pertaining to new. Should the definition of _DiSPATCHER_CONTEXT be ifdefed out on X86 to allow the conflicting one in clrnt.h to be used?

**Comment by sergiy-k** [comment]
@borgdylan Yes, correct.

**Comment by jkotas** [comment]
> Should the definition of _DiSPATCHER_CONTEXT be ifdefed out on X86 to allow the conflicting one in clrnt.h to be used?

Yes.

> I open a PR so someone on the team could review what I am doing?

If you have a set of clean good changes, we would be certainly interested in PR. We prefer smaller incremental PRs for this kind of work.

**Comment by borgdylan** [comment]
After doing that, I get errors in coreclr/src/pal/inc/pal.h mentioning that the dispatcher context is not available. Should I include PAL headers like clrnt.h within pal.h?

**Comment by jkotas** [comment]
pal.h has its own definition of PAL_DISPATCHER_CONTEXT. I do not see any uses of DISPATCHER_CONTEXT in pal.h. What is the error you are getting?

clrnt.h should not be included in pal.h. It will cause more problems that it would solve.

**Comment by borgdylan** [comment]
```
In file included from /var/www/git/coreclr/src/palrt/coguid.cpp:14:
In file included from /var/www/git/coreclr/src/palrt/common.h:17:
In file included from /var/www/git/coreclr/src/inc/winwrap.h:46:
In file included from /var/www/git/coreclr/src/inc/crtwrap.h:27:
In file included from /var/www/git/coreclr/src/pal/inc/rt/windows.h:14:
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1601:1: error: unknown type name 'PRUNTIME_FUNCTION'
PRUNTIME_FUNCTION
^
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1614:5: error: unknown type name 'PRUNTIME_FUNCTION'
    PRUNTIME_FUNCTION* Functions
    ^
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1640:5: error: /var/www/git/coreclr/src/pal/inc/rt/palrt.hunknown type name 'PRUNTIME_FUNCTION':
1640:5: error: unknown type name 'PRUNTIME_FUNCTION'
    PRUNTIME_FUNCTION FunctionEntry;
    ^
    PRUNTIME_FUNCTION FunctionEntry;
    ^
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1614:5: error: unknown type name 'PRUNTIME_FUNCTION'
    PRUNTIME_FUNCTION* Functions
    ^
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1640:5: error: unknown type name 'PRUNTIME_FUNCTION'
    PRUNTIME_FUNCTION FunctionEntry;
    ^
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1678:9: error: unknown type name 'DISPATCHER_CONTEXT'; did you mean 'PAL_DISPATCHER_CONTEXT'?
typedef DISPATCHER_CONTEXT *PDISPATCHER_CONTEXT;
        ^~~~~~~~~~~~~~~~~~
        PAL_DISPATCHER_CONTEXT
/var/www/git/coreclr/src/pal/inc/pal.h:5891:3: note: 'PAL_DISPATCHER_CONTEXT' declared here
} PAL_DISPATCHER_CONTEXT;
  ^
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1678:9: error: unknown type name 'DISPATCHER_CONTEXT'; did you mean 'PAL_DISPATCHER_CONTEXT'?
typedef DISPATCHER_CONTEXT *PDISPATCHER_CONTEXT;
        ^~~~~~~~~~~~~~~~~~
        PAL_DISPATCHER_CONTEXT
/var/www/git/coreclr/src/pal/inc/pal.h:5891:3: note: 'PAL_DISPATCHER_CONTEXT' declared here
} PAL_DISPATCHER_CONTEXT;
  ^
```

Also there are errors related to the printf function used n the PAL as commented before:
The PAL defines many printf variants using char * notation
While another file "printfcpp.hpp" defines them using LPSTR notation

The compiler is saying that the two definitions are not matching up in their types (conceptually they are the same definitions)?
We all know that on UNIX LPSTR becomes char * and LPCSTR becomes const char *. Should the code in printfcpp.hpp have its types change or should the PAL define LPSTR and LPCSTR ins terms of char * and const char *?

**Comment by janvorli** [comment]
@borgdylan Can you paste here the errors related to the LPSTR vs char*?

**Comment by borgdylan** [comment]
```
In file included from /var/www/git/coreclr/src/pal/src/arch/i386/context.cpp:24:
In file included from /var/www/git/coreclr/src/pal/src/include/pal/dbgmsg.h:162:
In file included from /var/www/git/coreclr/src/pal/src/include/pal/thread.hpp:33:
In file included from /var/www/git/coreclr/src/pal/src/include/pal/threadsusp.hpp:37:
/var/www/git/coreclr/src/pal/src/include/pal/printfcpp.hpp:37:5: error: conflicting types for 'PAL__vsnprintf'
    PAL__vsnprintf(
    ^
/var/www/git/coreclr/src/pal/inc/pal.h:5360:23: note: previous declaration is here
PALIMPORT int __cdecl _vsnprintf(char *, size_t, const char *, va_list);
                      ^
/var/www/git/coreclr/src/pal/inc/pal.h:5294:23: note: expanded from macro '_vsnprintf'
#define _vsnprintf    PAL__vsnprintf
                      ^
In file included from /var/www/git/coreclr/src/pal/src/arch/i386/context.cpp:24:
In file included from /var/www/git/coreclr/src/pal/src/include/pal/dbgmsg.h:162:
In file included from /var/www/git/coreclr/src/pal/src/include/pal/thread.hpp:33:
In file included from /var/www/git/coreclr/src/pal/src/include/pal/threadsusp.hpp:37:
/var/www/git/coreclr/src/pal/src/include/pal/printfcpp.hpp:45:5: error: conflicting types for 'PAL__wvsnprintf'
    PAL__wvsnprintf(
    ^
/var/www/git/coreclr/src/pal/inc/pal.h:5405:23: note: previous declaration is here
PALIMPORT int __cdecl _vsnwprintf(wchar_t *, size_t, const wchar_t *, va_list);
                      ^
/var/www/git/coreclr/src/pal/inc/pal.h:5295:23: note: expanded from macro '_vsnwprintf'
#define _vsnwprintf   PAL__wvsnprintf
                      ^
```

**Comment by jkotas** [comment]
>>> unknown type name 'PRUNTIME_FUNCTION'
>>> unknown type name 'DISPATCHER_CONTEXT'
To fix these errors, add "#if !defined(_TARGET_X86_)" ifdef around all instances of these types in palrt.h

**Comment by jkotas** [comment]
> unknown type name 'PRUNTIME_FUNCTION'
> unknown type name 'DISPATCHER_CONTEXT'

To fix these errors, add "#if !defined(_TARGET_X86_)" ifdef around all instances of these types in palrt.h

**Comment by jkotas** [comment]
> unknown type name 'PRUNTIME_FUNCTION'
> unknown type name 'DISPATCHER_CONTEXT'

To fix these errors, add `#if !defined(_TARGET_X86_)` ifdef around all instances of these types in palrt.h

**Comment by borgdylan** [comment]
After fixing these errors I get told that the KNONVOLATILE struct is not xietent:
```
In file included from /var/www/git/coreclr/src/md/ceefilegen/blobfetcher.cpp:12:
In file included from /var/www/git/coreclr/src/md/ceefilegen/stdafx.h:14:
In file included from /var/www/git/coreclr/src/inc/crtwrap.h:27:
In file included from /var/www/git/coreclr/src/pal/inc/rt/windows.h:14:
/var/www/git/coreclr/src/pal/inc/rt/palrt.h:1715:42: error: unknown type name 'KNONVOLATILE_CONTEXT_POINTERS'
BOOL PAL_VirtualUnwind(CONTEXT *context, KNONVOLATILE_CONTEXT_POINTERS *contextPointers);
```

**Comment by jkotas** [comment]
KNONVOLATILE_CONTEXT_POINTERS and its uses should be under  #if !defined(_TARGET_X86_)  as well.

**Comment by voltagex** [comment]
Should this issue be renamed and a separate issue opened for ARM(hf) platforms?

**Comment by gowrikumar** [comment]
@borgdylan, @jkotas 
May I know the status for this and how could I contribute towards bringing up on the Linux 32-bit.

**Comment by borgdylan** [comment]
You may fork my code. We do need direction from the coreclr team on important decisions however.

**Comment by kangaroo** [comment]
I think we've pretty severely diverged from the original issue.  If the focus of the goal is a 32-bit Linux port, which is not in the immediate interest of the project owners to my knowledge, this might be better served in a PR for code reviews, rather than living as an issue?

**Comment by MattWhilden** [comment]
It looks like people have enough info to get started. We're happy to pull fixes to support this into master but, as noted, this thread has become quite large.

Feel free to open smaller issues when they come up.

If you feel as thought I've closed this hastily, please let me know.
